Add API function to search for builds by component and version

mprpic opened the pull request:
FACTORY-2999

mprpic committed:
Add API function to search for builds by component and version

sdherr commented on assayist/client/query.py:
```diff
@@ -15,6 +15,31 @@ def set_connection(neo4j_url):
+        MATCH (c:Component {{canonical_name: '{name}'}}) <-[:SOURCE_FOR]-
+              (:SourceLocation {{canonical_version: '{version}'}}) <-[:UPSTREAM|EMBEDS*0..]-
+              (target_sl:SourceLocation) <-[:BUILT_FROM]- (build:Build)
+        OPTIONAL MATCH (target_sl) <-[:SUPERSEDES*0..]- (sl:SourceLocation)
```
You don't seem to be doing anything with this optional match. Unless I'm greatly mistaken, target_sl will remain what it was in the MATCH clause. You then optionally find sl matches that are superseeded by the target_sls, but never do anything with them. Also it naively looks to me like the direction on the supersedes relationship is backwards, but maybe I'm just crazy.

mprpic replied:
Should be fixed in current version. (And not, not crazy, totally correct ;-))

sdherr commented on assayist/client/query.py:
```diff
@@ -15,6 +15,31 @@ def set_connection(neo4j_url):
+              (:SourceLocation {{canonical_version: '{version}'}}) <-[:UPSTREAM|EMBEDS*0..]-
+              (target_sl:SourceLocation) <-[:BUILT_FROM]- (build:Build)
+        OPTIONAL MATCH (target_sl) <-[:SUPERSEDES*0..]- (sl:SourceLocation)
+        OPTIONAL MATCH (build) <-[:EMBEDS*]- (:Artifact) <-[:EMBEDS]- (target_sl)
```
Well first of all this will never match anything because Builds do not have an EMBEDS edge. You need an extra relationship there to link "build" with the embedded artifacts. Secondly it is not clear to me what this is trying to accomplish. I guess "Find additional builds that embed the source location"? If so I'm not sure this does what you'd expect, because "build" is already a defined set of nodes from the MATCH clause. I think this would use that list of Builds and look for relationships that match, but not expand the list (although I could be wrong). I would also appreciate line comments that state the intend of the clause.

mprpic replied:
Yep, the test case was also modelled incorrect to directly link build to the rsyslog source location. Fixed in current version.

sdherr commented on assayist/client/query.py:
```diff
@@ -15,6 +15,31 @@ def set_connection(neo4j_url):
+              (target_sl:SourceLocation) <-[:BUILT_FROM]- (build:Build)
+        OPTIONAL MATCH (target_sl) <-[:SUPERSEDES*0..]- (sl:SourceLocation)
+        OPTIONAL MATCH (build) <-[:EMBEDS*]- (:Artifact) <-[:EMBEDS]- (target_sl)
+        RETURN build
```
Unless I'm mistaken I think there's an addition case that's missing here: Builds that embed Artifacts that were built from the builds. Right now you're only trying to look for embedded Artifacts that embed the target_sl, not Artifacts that embed the produced Artifacts. Although maybe that's intentional and you're only looking for "first order" affected builds.

sdherr reviewed (changes requested):
A couple of thoughts:
1) It is not clear to me that just using the same variable name in several different MATCH clauses expands the list of results in that variable instead of restricting the nodes that will be match the clauses. If it does expand the list, then maybe some of my comments are invalid.

2) Line comments stating the intent of the clause are highly desirable so we can figure out what we were trying to do later.

3) We probably need to spend some time and map out test data for various complicated situations. That seems like the kind of thing that would be useful for testing various queries, and there's less chance of mistakes (both in the queries and in the tests) if we do it once independently instead of trying to figure it out for each query.

4) Maybe we need a cheat-sheet of common Cypher patterns.

Not all of those are on you to implement in this PR, but they are things that might be helpful here.

mprpic committed:
Add matching preceding versions

mprpic commented:
This is pretty much ready for review. There are individual tests that test specific scenarios. The graph setup for each should probably be moved to a factory class to unify it with other tests that use similar setups. I'll leave that for FACTORY-3364.

mprpic commented:
> A couple of thoughts:
> 
> 1. It is not clear to me that just using the same variable name in several different MATCH clauses expands the list of results in that variable instead of restricting the nodes that will be match the clauses. If it does expand the list, then maybe some of my comments are invalid.
> 

Query now uses proper variable names.

> 2. Line comments stating the intent of the clause are highly desirable so we can figure out what we were trying to do later.
> 

Added!

> 3. We probably need to spend some time and map out test data for various complicated situations. That seems like the kind of thing that would be useful for testing various queries, and there's less chance of mistakes (both in the queries and in the tests) if we do it once independently instead of trying to figure it out for each query.
> 

FACTORY-3364

> 4. Maybe we need a cheat-sheet of common Cypher patterns.
> 

FACTORY-3365

> 
> Not all of those are on you to implement in this PR, but they are things that might be helpful here.

twaugh commented on assayist/client/query.py:
```diff
@@ -15,6 +15,66 @@ def set_connection(neo4j_url):
+    :rtype: list
+    """
+    query = """
+        /// Find a component and a sourcelocation with the specified name and version. These can
```
Only need two forward-slashes for a line comment.

mprpic replied:
Fixed, thanks!

twaugh commented on assayist/client/query.py:
```diff
@@ -15,6 +15,66 @@ def set_connection(neo4j_url):
+        MATCH (c:Component {{canonical_name: '{name}'}}) <-[:SOURCE_FOR]-
+              (version_sl:SourceLocation {{canonical_version: '{version}'}})
+              <-[:UPSTREAM|EMBEDS*0..]- (target_sl:SourceLocation)
+              <-[:BUILT_FROM]- (direct_builds:Build)
```
If you put a `(:SourceLocation) <-[:SUPERSEDES*0..]-` before the `target_sl` node, won't you get all the builds you want without needing to use `WITH`?

mprpic replied:
Right, that would work if the `SUPERSEDES` relationship was defined between the dist-git SLs only since only those will then have a `BUILD_FROM` relationship with a build. I assumed here that `SUPERSEDES` would be between upstream SLs, which requires that extra hop from upstream SL to dist-git SL to build (if that makes sense).

If I rework the query to use `SUPERSEDES` on dist-git SLs, then yes, the `WITH` can be removed.

twaugh commented on assayist/client/query.py:
```diff
@@ -15,6 +15,66 @@ def set_connection(neo4j_url):
+        /// Find all container builds that embed an artifact produced by any of the previously
+        /// matched builds.
+        OPTIONAL MATCH (all_builds) -[:PRODUCED]-> (:Artifact) <-[:EMBEDS*0..]-
+                       (:Artifact) <-[:PRODUCED]- (container_build:Build {{type: 'container'}})
```
I think nodes in all_builds can't also be in container_build. Not sure if that's a problem.

mprpic replied:
Unless my thinking here is incorrect, that should be fine since this query is limited to returning container builds which will always embed the artifacts from builds that we are interested in. If this had to be generalized to RPM builds as well, `all_builds` would be a part of the results.

sdherr reviewed (approved):
You addressed my previous comments, looks much better.

mprpic committed:
Fix comment extra slashes in query comments

mprahl commented on assayist/client/query.py:
```diff
@@ -15,6 +15,66 @@ def set_connection(neo4j_url):
     neomodel.db.set_connection(neo4j_url)
 
 
+def get_container_by_component(component_name, component_version):
```
You should consider adding a parameter for the component type as well.

mprahl commented on assayist/client/query.py:
```diff
@@ -15,6 +15,66 @@ def set_connection(neo4j_url):
+        // used for a build.
+        MATCH (c:Component {{canonical_name: '{name}'}}) <-[:SOURCE_FOR]-
+              (version_sl:SourceLocation {{canonical_version: '{version}'}})
+              <-[:UPSTREAM|EMBEDS*0..]- (target_sl:SourceLocation)
```
Why is the upstream relationship used here?

mprahl commented on assayist/client/query.py:
```diff
@@ -15,6 +15,66 @@ def set_connection(neo4j_url):
+        MATCH (c:Component {{canonical_name: '{name}'}}) <-[:SOURCE_FOR]-
+              (version_sl:SourceLocation {{canonical_version: '{version}'}})
+              <-[:UPSTREAM|EMBEDS*0..]- (target_sl:SourceLocation)
+              <-[:BUILT_FROM]- (direct_builds:Build)
```
Since `direct_builds` isn't a collection, could you make it singular?

mprahl commented on assayist/client/query.py:
```diff
@@ -15,6 +15,66 @@ def set_connection(neo4j_url):
+        WITH COLLECT(target_sl) + COLLECT(older_target_sl) AS sls, builds
+
+        // Expand lists into rows so we can match on them.
+        UNWIND builds as all_builds
```
This is no longer a collection, so could you make `all_builds` singular?

mprahl commented on assayist/client/query.py:
```diff
@@ -15,6 +15,66 @@ def set_connection(neo4j_url):
+
+        // Expand lists into rows so we can match on them.
+        UNWIND builds as all_builds
+        UNWIND sls as all_sls
```
This is no longer a collection, so could you make `all_sls` singular?

mprahl commented on assayist/client/query.py:
```diff
@@ -15,6 +15,66 @@ def set_connection(neo4j_url):
+        // Find a component and a sourcelocation with the specified name and version. These can
+        // be emdedded in other sourcelocation that are used as a source for a build or directly
+        // used for a build.
+        MATCH (c:Component {{canonical_name: '{name}'}}) <-[:SOURCE_FOR]-
```
Since this is a `MATCH` and not an `OPTIONAL MATCH`, the whole query will return nothing if none of the nodes in `target_sl` have a build associated with it. Based on the supersedes optional match below, I don't think that's what you want. I suggest splitting this up in two queries (untested):

```cypher
 MATCH (c:Component {{canonical_name: '{name}'}}) <-[:SOURCE_FOR]-version_sl:SourceLocation {{canonical_version: '{version}'}})<-[:UPSTREAM|EMBEDS*0..]- (target_sl:SourceLocation)

OPTIONAL MATCH (target_sl)<-[:BUILT_FROM]- (direct_builds:Build)
```

mprahl commented on assayist/client/query.py:
```diff
@@ -15,6 +15,66 @@ def set_connection(neo4j_url):
+    for result in results:
+        for node in result:
+            if node:
+                build_ids.add(node['id'])
```
This is a string in Neo4j, but you may want to return these as integers.

mprahl commented on tests/client/test_query.py:
```diff
@@ -2,7 +2,271 @@
+
+    rpm_component = Component(canonical_name='rsyslog', canonical_type='generic',
+                              canonical_namespace='').save()
+    rpm_upstream_source.component.connect(rpm_component)
```
The internal source should also connect with the component.

mprahl commented on tests/client/test_query.py:
```diff
@@ -2,7 +2,271 @@
+    preceding_rpm_internal_source_url = (
+        'git://pkgs.devel.redhat.com/rpms/rsyslog#008741c08b950e5d60620dae44c8eabd78391706'
+    )
+    preceding_rpm_internal_source = SourceLocation(url=preceding_rpm_internal_source_url).save()
```
Why isn't there a `preceding_rpm_internal_source.previous_version.connect(rpm_internal_source)`?

mprahl commented on tests/client/test_query.py:
```diff
@@ -2,7 +2,271 @@
+
+def test_get_container_by_embedded_component():
+    """Test the get_container_by_component function with a container that embeds a component."""
+    # Set up a layered container image build, build, whose single
```
What do you mean by `image build, build,`?

mprahl commented on tests/client/test_query.py:
```diff
@@ -2,7 +2,271 @@
+    bash_internal_source = SourceLocation(url=bash_internal_url).save()
+    bash_build.source_location.connect(bash_internal_source)
+    bash_internal_source.upstream.connect(bash_upstream_source)
+    bash_component = Component(canonical_type='generic', canonical_namespace='',
```
Why is `canonical_namespace` an empty string and not `None`/omitted?

mprpic replied:
Because it's a required attribute, which leads me to the question: does it have to be required? I bumped into this in #26 as well.

mprahl replied:
@mprpic could you please bring this up in stand up?

mprahl reviewed (changes requested):
Just a few things to address. Please let me know once you are ready for review again.

mprpic committed:
Update query to disregard preceding versions

These will be queried separately with `get_current_and_previous_versions`
API query from PR #17.

mprpic committed:
Merge branch 'master' into api-search-by-component

Conflicts:
        tests/client/test_query.py

mprpic commented:
> Just a few things to address. Please let me know once you are ready for review again.

@mprahl Thanks for the feedback! Ready for re-review.

mprpic committed:
Merge branch 'master' into api-search-by-component

Conflicts:
        tests/client/test_query.py

mprahl commented on assayist/client/query.py:
```diff
@@ -18,6 +18,51 @@ def set_connection(neo4j_url):
+        // used for a build.
+        MATCH (c:Component {{canonical_name: '{name}', canonical_type: '{type}'}}) <-[:SOURCE_FOR]-
+              (version_sl:SourceLocation {{canonical_version: '{version}'}})
+              <-[:UPSTREAM|EMBEDS*0..]- (target_sl:SourceLocation)
```
So if you pass in an internal only canonical version then all the source locations that embed it will be found. If you pass in an upstream only canonical version, then the embeds will have no effect because the upstream source locations will not have embeds. Is this desired behavior?

Perhaps what you want to do is:
```
MATCH (c:Component {{canonical_name: '{name}', canonical_type: '{type}'}}) <-[:SOURCE_FOR]-
    (version_sl:SourceLocation {{canonical_version: '{version}'}})
    <-[:UPSTREAM*0..]-(:SourceLocation)<-[:EMBEDS*0..]-(target_sl:SourceLocation)
```

mprahl commented:
@mprpic I left one comment. Let me know if that makes sense.

mprpic committed:
Fix incorrect cardinality on upstream relationship

mprahl commented:
@mprpic in the future, could you rebase instead of merging master into your branch, and could you squash the commits per change/feature?

mprpic commented:
> @mprpic in the future, could you rebase instead of merging master into your branch, and could you squash the commits per change/feature?

@mprahl Rebasing will add loads of commits onto the branch that then make reviewing the actual changes difficult. Actually, that's how it works in GitLab, not sure about GitHub so correct me if my assumption is wrong. If I include it as a single merge commit, it is ignored in the diff of all changes. (**Edit**: ok, nevermind, I just rebased master in #26 and it ignores those commits in the PR diff.)

As for squashing, GitHub gives you the option to squash before merging which I find more useful than constantly squashing all commits into one during the lifetime of the PR. It's sometimes nice to view just the changes applied after some commit that fixes things from reviewers' comments.
